### ISSUE TITLE
Support soft-wrapping for code blocks

### ISSUE THREAD

**Issue by damithc** [MarkBind/markbind#2404]
For cases like the below, can we provide some sort of soft-wrapping support (similar to Intellij IDEA) so that the hidden part of the text can be made visible, especially for printing?

![image](https://github.com/MarkBind/markbind/assets/1673303/8d91affc-265b-423b-a5d6-217a7703e60d)

**Comment by Tim-Siu** [comment]
Adding Intellij IDEA like indention-aware code wrapping can be quite hard. A pure CSS approach may not be viable due to the nature of how whitespace is handled in HTML and CSS. Current logic in markdown-it does not involve indentation. We might have to work on the parser there.

A naive soft wrapping preserves the original line breaks and spaces but does not add additional indentation to each wrapped line.
<img width="995" alt="Screenshot 2024-02-05 at 13 01 46" src="https://github.com/MarkBind/markbind/assets/61866948/ea22ed2b-4f5b-4284-8c2d-6b52368228a9">

Should we consider having this naive soft-wrapping in printing first?

**Comment by yucheng11122017** [comment]
Hi @damithc, there is some existing functionality already with the plugin [CodeWrap](https://markbind.org/userGuide/usingPlugins.html#plugin-codeblockwrapbuttons).
If the user has already wrapped the text using the plugin, when printing the wrap is maintained (ie the code is wrapped) 
![image](https://github.com/MarkBind/markbind/assets/77204346/282000bc-8335-4969-ad89-69d00076f7db)

Otherwise, it is as shown in your display. 
Would it be sufficient to make wrapping text the default behaviour when this is plugin is enabled so that when users print, it should already be wrapped unless chosen to do so? 
Or should be code will be wrapped as long as there is printing? (regardless of whether this plugin is enabled and whether or not the wrap text button was clicked)

**Comment by damithc** [comment]
> Hi @damithc, there is some existing functionality already with the plugin [CodeWrap](https://markbind.org/userGuide/usingPlugins.html#plugin-codeblockwrapbuttons). If the user has already wrapped the text using the plugin, when printing the wrap is maintained (ie the code is wrapped) ![image](https://private-user-images.githubusercontent.com/77204346/303893905-282000bc-8335-4969-ad89-69d00076f7db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc2MjY5ODYsIm5iZiI6MTcwNzYyNjY4NiwicGF0aCI6Ii83NzIwNDM0Ni8zMDM4OTM5MDUtMjgyMDAwYmMtODMzNS00OTY5LWFkODktNjlkMDAwNzZmN2RiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjExVDA0NDQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1Njc0ODI3NmE3NjNhNjEzZGU5ZTNiMjU5MzY3MGZiZWZhMTFhODY3MTEzYTZiMjY5YWMzZGYwNGU1Y2E0NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RyUabnk1aTD4WnZnBtIv_67yyjItmfGsIARndfHwI5s)
> 
> Otherwise, it is as shown in your display. Would it be sufficient to make wrapping text the default behaviour when this is plugin is enabled so that when users print, it should already be wrapped unless chosen to do so? Or should be code will be wrapped as long as there is printing? (regardless of whether this plugin is enabled and whether or not the wrap text button was clicked)

Thanks for the investigation @yucheng11122017 

1. I feel auto-wrapping should be done for printing no matter what. What do you think?
1. Should we enable the plugin by default? Seems like a useful functionality to have.
2. Not sure if wrapping should be the default behavior for normal viewing, as wrapped code may not be syntactically same as  the original code, for space-sensitive languages such as Python.
3. Related to the above, the code wrapping behavior of the plugin is affected by this flaw as well. It is much better if there is a way to subtly differentiate soft line breaks that we inserted from actual line breaks in the original text.

**Comment by Tim-Siu** [comment]
I propose modifying `markdown-it.js` to wrap each line of code from the original markdown in an HTML element with a distinct CSS class. When processing each line of a code block, calculate the indentation level (e.g., the number of spaces or tabs at the beginning of the line) and set it as a CSS variable on the element wrapping that line. We can utilize CSS custom properties to dynamically adjust the indentation based on the content.

Sample HTML output from `markdown-it.js` (irrelevant details ignored):
```html
<pre>
<span class="code-line" style="--indent-size: 0px;">function example() {</span>
<span class="code-line" style="--indent-size: 20px;">console.log("Hello, world!");</span>
<span class="code-line" style="--indent-size: 0px;">}</span>
</pre>
```

Example CSS: 
```css
.code-line {
  display: inline-block;
  white-space: pre-wrap; /* Allow wrapping */
  --indent-size: 0px; /* Default indentation size */
  margin-left: var(--indent-size); /* Apply indentation */
}
```

**Comment by yucheng11122017** [comment]
> I propose modifying `markdown-it` to wrap each line of code from the original markdown in an HTML element with a distinct CSS class. When processing each line of a code block, calculate the indentation level (e.g., the number of spaces or tabs at the beginning of the line) and set it as a CSS variable on the element wrapping that line. We can utilize CSS custom properties to dynamically adjust the indentation based on the content.
> 
> Sample HTML output from `markdown-it` (irrelevant details ignored):
> 
> ```
> <pre>
> <span class="code-line" style="--indent-size: 0px;">function example() {</span>
> <span class="code-line" style="--indent-size: 20px;">console.log("Hello, world!");</span>
> <span class="code-line" style="--indent-size: 0px;">}</span>
> </pre>
> ```
> 
> Example CSS:
> 
> ```css
> .code-line {
>   display: inline-block;
>   white-space: pre-wrap; /* Allow wrapping */
>   --indent-size: 0px; /* Default indentation size */
>   margin-left: var(--indent-size); /* Apply indentation */
> }
> ```

Hi @Tim-Siu have you experimented with this? I'm wondering how accurate will the calculation be of the indentation..

**Comment by damithc** [comment]
> 3\. ot sure if wrapping should be the default behavior for normal viewing, as wrapped code may not be syntactically same as  the original code, for space-sensitive languages such as Python.

Regarding this, I just realized that if line numbers are shown, there is an indirect indication which lines have been soft-wrapped. Perhaps we can use that? i.e., show line numbers when soft-wrapping, even if line numbers is not enabled for that code block

![image](https://github.com/MarkBind/markbind/assets/1673303/51351617-2e86-4a75-9bb7-32ab4791cc06)

**Comment by Tim-Siu** [comment]
> > ... as soft-wrapped code may not be syntactically same as  the original code, for space-sensitive languages such as Python.
> 
> Regarding this, I just realized that if line numbers are shown, there is an indirect indication which lines have been soft-wrapped. Perhaps we can use that? i.e., show line numbers when soft-wrapping, even if line numbers is not enabled for that code block
> 
> ![image](https://private-user-images.githubusercontent.com/1673303/303902220-51351617-2e86-4a75-9bb7-32ab4791cc06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc2MzM2NTAsIm5iZiI6MTcwNzYzMzM1MCwicGF0aCI6Ii8xNjczMzAzLzMwMzkwMjIyMC01MTM1MTYxNy0yZTg2LTRhNzUtOWJiNy0zMmFiNDc5MWNjMDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTFUMDYzNTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGZjMDUzZTQzYThkN2FhZmE1NDZmOWRlMzQzOWY0OWIxNjdmYmVjNTU3NWY3MmE4ZjBlYmZlY2VmZWVlZWJlNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.P2gEgCOC4C7zmHmoCPF8dMU3ty1TZbBYYN7BAiElzns)

Thank you for the observation. This indeed appears to be the perfect solution. To move forward with this, we can:

1. Make code soft-wrapping the bottom default setting, instead of offering it merely as a plugin. This ensures all users automatically benefit from better readability without the need for manual activation.
2. Apply code soft-wrapping when printing, ensuring that the printed is complete.
3. Automatically use line numbers with soft-wrapped code, even if not specifically enabled by the user. 

I will raise a PR to implement these changes.

**Comment by yucheng11122017** [comment]
> 1. Make code soft-wrapping the bottom default setting, instead of offering it merely as a plugin. This ensures all users automatically benefit from better readability without the need for manual activation.
> 2. Apply code soft-wrapping when printing, ensuring that the printed is complete.
> 3. Automatically use line numbers with soft-wrapped code, even if not specifically enabled by the user.
> 

Hi @Tim-Siu, the latter two sound good. 
I'm not too sure about the first point though - whether we would want code wrapping to be the default setting instead of a plugin. Maybe anyone else would care to input? 
In the meantime, you can do a PR for the latter two first that would be great!

**Comment by yucheng11122017** [comment]
Closed by #2431
